File: greenfloor-engine/src/storage/sqlite/coin_ops.rs
use rusqlite::params;

use crate::error::{SignerError, SignerResult};

use super::{utcnow_iso, SqliteStore};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CoinOpBudgetReport {
    pub spent_mojos: i64,
    pub executed_ops: i64,
    pub planned_ops: i64,
    pub skipped_ops: i64,
    pub fee_budget_skipped_ops: i64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CoinOpLedgerEntry<'a> {
    pub market_id: &'a str,
    pub op_type: &'a str,
    pub op_count: i64,
    pub fee_mojos: i64,
    pub status: &'a str,
    pub reason: &'a str,
    pub operation_id: Option<&'a str>,
}

impl SqliteStore {
    pub fn get_daily_fee_spent_mojos_utc(&self) -> SignerResult<i64> {
        let total: i64 = self
            .conn
            .query_row(
                r#"
                SELECT COALESCE(SUM(fee_mojos), 0)
                FROM coin_op_ledger
                WHERE date(created_at) = date('now')
                  AND status = 'executed'
                "#,
                [],
                |row| row.get(0),
            )
            .map_err(|err| {
                SignerError::Other(format!("failed to query daily coin-op fee total: {err}"))
            })?;
        Ok(total)
    }

    pub fn add_coin_op_ledger_entry(&self, entry: CoinOpLedgerEntry<'_>) -> SignerResult<()> {
        let CoinOpLedgerEntry {
            market_id,
            op_type,
            op_count,
            fee_mojos,
            status,
            reason,
            operation_id,
        } = entry;
        self.conn
            .execute(
                r#"
                INSERT INTO coin_op_ledger
                  (market_id, op_type, op_count, fee_mojos, status, reason, operation_id, created_at)
                VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8)
                "#,
                params![
                    market_id,
                    op_type,
                    op_count,
                    fee_mojos,
                    status,
                    reason,
                    operation_id,
                    utcnow_iso(),
                ],
            )
            .map_err(|err| {
                SignerError::Other(format!("failed to insert coin_op_ledger row: {err}"))
            })?;
        Ok(())
    }

    pub fn get_coin_op_budget_report_utc(&self) -> SignerResult<CoinOpBudgetReport> {
        self.conn
            .query_row(
                r#"
                SELECT
                  COALESCE(SUM(CASE WHEN status = 'executed' THEN fee_mojos ELSE 0 END), 0),
                  COALESCE(SUM(CASE WHEN status = 'executed' THEN op_count ELSE 0 END), 0),
                  COALESCE(SUM(CASE WHEN status = 'planned' THEN op_count ELSE 0 END), 0),
                  COALESCE(SUM(CASE WHEN status = 'skipped' THEN op_count ELSE 0 END), 0),
                  COALESCE(SUM(CASE WHEN status = 'skipped' AND reason = 'fee_budget_guard' THEN op_count ELSE 0 END), 0)
                FROM coin_op_ledger
                WHERE date(created_at) = date('now')
                "#,
                [],
                |row| {
                    Ok(CoinOpBudgetReport {
                        spent_mojos: row.get(0)?,
                        executed_ops: row.get(1)?,
                        planned_ops: row.get(2)?,
                        skipped_ops: row.get(3)?,
                        fee_budget_skipped_ops: row.get(4)?,
                    })
                },
            )
            .map_err(|err| {
                SignerError::Other(format!("failed to query coin-op budget report: {err}"))
            })
    }
}

File: tests/helpers/sqlite_store/coin_ops.py
from __future__ import annotations

from tests.helpers.sqlite_store.base import SqliteStoreMixin
from tests.helpers.sqlite_store.schema import utcnow_iso


class CoinOpStoreMixin(SqliteStoreMixin):
    def add_coin_op_ledger_entry(
        self,
        *,
        market_id: str,
        op_type: str,
        op_count: int,
        fee_mojos: int,
        status: str,
        reason: str,
        operation_id: str | None,
    ) -> None:
        self.conn.execute(
            """
            INSERT INTO coin_op_ledger
              (market_id, op_type, op_count, fee_mojos, status, reason, operation_id, created_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                market_id,
                op_type,
                int(op_count),
                int(fee_mojos),
                status,
                reason,
                operation_id,
                utcnow_iso(),
            ),
        )
        self.conn.commit()

    def get_daily_fee_spent_mojos_utc(self) -> int:
        row = self.conn.execute(
            """
            SELECT COALESCE(SUM(fee_mojos), 0) AS total
            FROM coin_op_ledger
            WHERE date(created_at) = date('now')
              AND status = 'executed'
            """
        ).fetchone()
        if row is None:
            return 0
        return int(row["total"] or 0)

    def get_coin_op_budget_report_utc(self) -> dict:
        row = self.conn.execute(
            """
            SELECT
              COALESCE(SUM(CASE WHEN status = 'executed' THEN fee_mojos ELSE 0 END), 0) AS spent_mojos,
              COALESCE(SUM(CASE WHEN status = 'executed' THEN op_count ELSE 0 END), 0) AS executed_ops,
              COALESCE(SUM(CASE WHEN status = 'planned' THEN op_count ELSE 0 END), 0) AS planned_ops,
              COALESCE(SUM(CASE WHEN status = 'skipped' THEN op_count ELSE 0 END), 0) AS skipped_ops,
              COALESCE(SUM(CASE WHEN status = 'skipped' AND reason = 'fee_budget_guard' THEN op_count ELSE 0 END), 0) AS fee_budget_skipped_ops
            FROM coin_op_ledger
            WHERE date(created_at) = date('now')
            """
        ).fetchone()
        if row is None:
            return {
                "spent_mojos": 0,
                "executed_ops": 0,
                "planned_ops": 0,
                "skipped_ops": 0,
                "fee_budget_skipped_ops": 0,
            }
        return {
            "spent_mojos": int(row["spent_mojos"] or 0),
            "executed_ops": int(row["executed_ops"] or 0),
            "planned_ops": int(row["planned_ops"] or 0),
            "skipped_ops": int(row["skipped_ops"] or 0),
            "fee_budget_skipped_ops": int(row["fee_budget_skipped_ops"] or 0),
        }
